### PR TITLE
[#2322] Fix various types for service with one offer

### LINKED ISSUE #2322: Resource with one offer is Various order type
**Describe the bug**
Tested on v3.22.0-beta.0"

**To Reproduce**
1. Go to the resource where you have administrator rights - Order type: Open Access
2. Delete all offers (via API)
3. Click the button Manage resource --> Set parameters and offers
image
4. Click the button "Add new offer"
5. Fill the form, field order type is editable, change to Order required, and save
6. See error -> resource is Various type although has one offer (resource is open access, offer order required)
image
7. Update resource in PC
8. See error - resource has two offers
image

Expected result:
offer is of the same type as resource

**similar problem**
1. Go to Resource with order type Open Access
2. Create second offer Order required 
3. Resource is Various Order type
image
4. Click "Edit" for every offer and open the link in the new tab
5. In one tab click "Delete offer"
6. In the second tab change the offer order type to Full Open Access and Update offer
7. See error - resource is various order type with one offer
image

### PR DESCRIPTION
**Pull request by goreck888**
Fixes posibility of setting different order types
for offer and resource if that resource has
only one offer
Fixes #2322

### CODE REVIEW

**Comment by Marcelinna** [comment]
1. Create new resource, fill only required fields? Why offer doesnt created?
image

**Comment by Marcelinna** [comment]
1. Go to PC
2. Create resource with order type: open access
3. Update resource in PC, change order type to full open access
In MP we have resource  Various order types  instead Full open Access
image

File: app/models/offer.rb [hunk 1]
```diff
@@ -100,6 +104,12 @@ def oms_params_match?
     end
   end
 
+  def same_order_type_as_in_service
+    unless order_type == service.order_type
+      errors.add(:order_type, "must be the same as in the resource: #{service.order_type}")
+    end
+  end
+
```
Comment (Marcelinna) [inline app/models/offer.rb:112]: see this bug too https://github.com/cyfronet-fid/marketplace/issues/2302

**Comment by Marcelinna** [comment]
If we don't fill field order_type in backoffce, whether we set to field by default?

File: app/models/offer.rb [hunk 2]
```diff
@@ -57,6 +57,14 @@ def should_index?
   validate :proper_oms?, if: -> { primary_oms.present? }
   validates :oms_params, absence: true, if: -> { current_oms.blank? }
   validate :check_oms_params, if: -> { current_oms.present? }
+  validate :same_order_type_as_in_service,
+           if: -> {
+             service&.order_type&.present? &&
+               (
+                 (service&.offers&.published&.size&.zero? && new_record?) ||
+                   (service&.offers&.published&.size == 1 && persisted?)
+               )
```
Comment (jswk) [inline app/models/offer.rb:66]: Most of the safe navigation operators are redundant here, this should suffice (existence of `service` is checked in the first line, `.present?` doesn't need the safe navigation):
```ruby
             service&.order_type.present? &&
               (
                 (new_record? && service.offers.published.size.zero?) ||
                   (persisted? && service.offers.published.size == 1)
               )
```
I would also put the record state check before offer counts. Seems easier to read.

File: app/models/service.rb [hunk 3]
```diff
@@ -39,7 +39,7 @@ class Service < ApplicationRecord
 
   enum status: STATUSES
 
-  has_many :offers, dependent: :restrict_with_error
+  has_many :offers, dependent: :restrict_with_error, validate: false
```
Comment (jswk) [inline app/models/service.rb:42]: I don't think it's needed. Setting `validate: false` here only disables validation for offers added to the relation.
The default of true should work ok.

File: app/views/backoffice/services/offers/_one_offer_edit_form.html.haml [hunk 4]
```diff
@@ -4,6 +4,7 @@
   .col-lg-8.pl-0{ "data-controller": offer.order_required? ? "ordering" : "" }
     = f.error_notification
     = f.hidden_field :id
+    = f.hidden_field :order_type, value: service.order_type
```
Comment (jswk) [inline app/views/backoffice/services/offers/_one_offer_edit_form.html.haml:7]: Why it's necessary here? Wouldn't a single offer already have a correct order_type set when editing it?

**Review by jswk — COMMENTED** [review]
It seems to work ok, I have only given a few minor correction ideas.
Also, please clarify the reason for the hidden field and disabling validation in relation.

**Comment by Marcelinna** [comment]
1. Go to Backoffice
2. Create resource order type: Full open access
3. Add second offer: open access
4. Add offer in two separate tab
5. Delete open access offer
6. For the second offer change open-access type to order requirder
7. Click button update
8. See problem - after first update user see view for order required defauilt offer
image
9. Click update - offer is update to full open access - ok

File: app/models/service.rb (hunk 3 above)
Comment (goreck888) [reply app/models/service.rb]: Done.

File: app/views/backoffice/services/offers/_one_offer_edit_form.html.haml (hunk 4 above)
Comment (goreck888) [reply app/views/backoffice/services/offers/_one_offer_edit_form.html.haml:7]: I think it's necessary. See @Marcelinna comment below, that's the reason of this addition.

File: app/models/offer.rb (hunk 2 above)
Comment (goreck888) [reply app/models/offer.rb]: Done.

**Review by jswk — APPROVED** [review]

**Review by jswk — APPROVED** [review]

**Review by Marcelinna — APPROVED** [review]

**Pull request merged**